components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,382 @@
+        )
+
+        val shellTerminalWidget =
+                terminalView.widgets.find { widget ->
+                    terminalView.toolWindow.contentManager.getContent(widget).tabName ==
+                    terminalRunnerId
+                } as ShellTerminalWidget
+
+        backendTerminalManager.shareTerminal(shellTerminalWidget, terminalRunnerId)
+
+        val terminalContent = terminalView.toolWindow.contentManager.getContent(shellTerminalWidget)
+
+        terminalContent.putUserData(TITLE_KEY, supervisorTerminal.title)
+
+        connectSupervisorStream(
+                shellTerminalWidget,
+                supervisorTerminal,
+                terminalOutputWriter,
+                terminalInputReader
+        )
+    }
+
+    private fun connectSupervisorStream(
+            shellTerminalWidget: ShellTerminalWidget,
+            supervisorTerminal: TerminalOuterClass.Terminal,
+            terminalOutputWriter: PipedOutputStream,
+            terminalInputReader: PipedInputStream
+    ) {
+        val dataReceivedFromSupervisor = ByteArrayOutputStream()
+
+        val listenTerminalRequest =
+                TerminalOuterClass.ListenTerminalRequest.newBuilder()
+                        .setAlias(supervisorTerminal.alias)
+                        .build()
+
+        val listenTerminalResponseObserver =
+                object : StreamObserver<TerminalOuterClass.ListenTerminalResponse> {
+                    override fun onNext(response: TerminalOuterClass.ListenTerminalResponse) {
+                        when {
+                            response.hasTitle() -> {
+                                debug("Received terminal title: ${response.title}")
+                                val terminalContent = terminalView.toolWindow.contentManager.getContent(shellTerminalWidget)
+                                terminalContent.putUserData(TITLE_KEY, response.title)
+                            }
+                            response.hasData() -> {
+                                debug("Printing a text on '${supervisorTerminal.title}' terminal.")
+                                dataReceivedFromSupervisor.write(response.data.toByteArray())
+                            }
+                            response.hasExitCode() -> {
+                                debug(
+                                        "Closing '${supervisorTerminal.title}' terminal (Exit Code: ${response.exitCode}."
+                                )
+                                shellTerminalWidget.close()
+                            }
+                        }
+                    }
+
+                    override fun onCompleted() {
+                        debug("'${supervisorTerminal.title}' terminal finished reading stream.")
+                    }
+
+                    override fun onError(throwable: Throwable) {
+                        if (containsTerminatedStatus(throwable)) return
```

Comment:
we should handle reconnections here, looks at how notifications are handled in `GitpodManager`